File: 0501-1000/0981. Time Based Key-Value Store.py
# Medium
# Map, Binary search
# https://leetcode.cn/problems/time-based-key-value-store/

class TimeMap:

    def __init__(self):
        self.store = {} # {key: [(timestamp, value), ...]}
    
    # Time: O(1)
    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.store:
            self.store[key] = []
        self.store[key].append((timestamp, value))

    # Time: O(log n)
    def get(self, key: str, timestamp: int) -> str:
        if key not in self.store:
            return ""
        
        arr = self.store[key]
        # use binary search to find the max val <= timestamp
        left, right = 0, len(arr) - 1
        res = ""

        while left <= right:
            mid = (left + right) // 2
            if arr[mid][0] <= timestamp:
                res = arr[mid][1] # update res
                left = mid + 1
            else:
                right = mid - 1

        return res

File: 0501-1000/0981. Time Based Key-Value Store.go
package main
// Medium
// HashMap, Binary Search
// https://leetcode.cn/problems/time-based-key-value-store/

type TimeMap struct {
    store map[string][]Pair
}

type Pair struct {
    timestamp int
    value string
}

func Constructor() TimeMap {
   return TimeMap{
    store: make(map[string][]Pair),
   } 
}

// Time:O(1), Space:O(n)
func (this *TimeMap) Set(key string, value string, timestamp int)  {
    this.store[key] = append(this.store[key], Pair {
        timestamp: timestamp,
        value: value,
    })
}

// Time:O(logm), Space:O(n)
func (this *TimeMap) Get(key string, timestamp int) string {
    pairs, exists := this.store[key]
    if !exists {
        return ""
    }

    // find the last pos that ts <= target
    left, right := 0, len(pairs) - 1
    res := -1

    for left <= right {
        mid := left + (right - left) / 2
        if pairs[mid].timestamp <= timestamp {
            // find a match, record and keep finding
            res = mid
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    if res == -1 {
        return ""
    }
    return pairs[res].value
}
